### PR TITLE
DAOS-7170 engine: avoid using XS0 for IO forwarding

### PR DESCRIPTION
**Pull request by liuxuezhao**
sched_ult2xs() should avoid using XS0 for IO forwarding, to avoid
blocking XS0.

Signed-off-by: Xuezhao Liu <xuezhao.liu@intel.com>

### CODE REVIEW

**Review by Nasf-Fan — DISMISSED** [review]

File: src/engine/ult.c [hunk 1]
```diff
@@ -304,13 +304,22 @@ dss_thread_collective(int (*func)(void *), void *arg, unsigned int flags)
...
 		return DSS_XS_SELF;
 	case DSS_XS_IOFW:
-		if (!dss_helper_pool)
-			return (DSS_MAIN_XS_ID(tgt_id) + 1) % DSS_XS_NR_TOTAL;
+		if (!dss_helper_pool) {
+			if (dss_tgt_offload_xs_nr > 0)
+				xs_id = DSS_MAIN_XS_ID(tgt_id) + 1;
+			else
+				xs_id = DSS_MAIN_XS_ID((tgt_id + 1) %
+						       dss_tgt_nr);
```
Comment (NiuYawei) [inline src/engine/ult.c:319]: I don't see why the calculated result of xs_id can't be 0 or 1 here.

File: src/engine/ult.c [hunk 2]
```diff
@@ -304,13 +304,22 @@ dss_thread_collective(int (*func)(void *), void *arg, unsigned int flags)
...
+			D_ASSERT(xs_id < DSS_XS_NR_TOTAL && xs_id > 0);
```
Comment (NiuYawei) [inline src/engine/ult.c:320]: xs_id < DSS_XS_NR_TOTAL && xs_id > 1

Comment (liuxuezhao) [reply src/engine/ult.c:320]: OK, I can refresh to restrict it better.

**Review by NiuYawei — DISMISSED** [review]

**Review by Nasf-Fan — DISMISSED** [review]

**Comment by liuxuezhao** [comment]
just a rebase, thx

**Comment by liuxuezhao** [comment]
@NiuYawei @Nasf-Fan just a rebase, thanks

**Review by NiuYawei — APPROVED** [review]

**Review by wangdi1 — APPROVED** [review]

**Pull request merged**